### PR TITLE
fix: Add helpful error message if wrong table is used for Where or OrderBy expression.

### PR DESCRIPTION
**Pull request by SandPod**
Closes: https://github.com/serverpod/serverpod/issues/1312

Adds a more helpful error message if a "Where" or "OrderBy" column expression is built from another table than the one we are performing operations on.

### Where expression error example:
``` dart
Citizen.find(
  session,
  where: (_) =>
      Company.t.name.equals('Serverpod'),
);
```
Will give an exception with message: 
```
FormatException: Column references starting from other tables than "citizen" are not supported. The following expressions need to be removed or modified:
"where" expression referencing column company."name".
```

### OrderBy expression error example:
``` dart
Citizen.find(
  session,
  orderBy: Company.t.name,
);
```

Will give an exception with message: 
```
FormatException: Column references starting from other tables than "citizen" are not supported. The following expressions need to be removed or modified:
"orderBy" expression referencing column company."name".
```

## Pre-launch Checklist

- [ ] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [ ] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [ ] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [ ] All existing and new tests are passing.
- [ ] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

-

### CODE REVIEW

File: packages/serverpod/lib/src/database/database_query.dart [hunk 1]
```diff
@@ -349,3 +355,42 @@ _UsingQuery _usingQueryFromTableRelations(
...
+        'Column references starting from other tables than "$tableName" are not supported. The following expressions need to be removed or modified:'
+        '\n${exceptions.join('\n')}';
+    throw FormatException(errorMessage);
+  }
+}
+
+extension _ColumnHelpers on Column {
+  /// Returns true if the column has the specified table as base table.
+  bool hasBaseTable(String table) {
+    // Regex matches '<tableName>_' and '<tableName>.'
```
Comment (Isakdl) [inline packages/serverpod/lib/src/database/database_query.dart:393]: Suggestion: The regex does not include the `<>` here I think it makes it more clear to not include chars that is not strictly matched in the regex. I.E. `'tableName_'` and `'tableName'`

File: packages/serverpod/test/database/database_query_test.dart [hunk 2]
```diff
@@ -53,28 +53,40 @@ void main() {
     test(
         'when query with single order by is built then output is single order by query.',
         () {
-      Order order = Order(column: ColumnString('id'), orderDescending: false);
+      var table = 'citizen';
+      Order order = Order(
+          column: ColumnString('id', queryPrefix: table),
+          orderDescending: false);
```
Comment (Isakdl) [inline packages/serverpod/test/database/database_query_test.dart:59]: NIT: add a `,` at the end

**Review by Isakdl — APPROVED** [review]
LGTM! With NIT changes! 🚀

File: packages/serverpod/lib/src/database/database_query.dart [hunk 3]
```diff
@@ -349,3 +355,42 @@ _UsingQuery _usingQueryFromTableRelations(
...
+    for (var column in columns) {
+      if (!column.hasBaseTable(tableName)) {
+        exceptions.add('"where" expression referencing column $column.');
+      }
+    }
+  }
+
+  if (exceptions.isNotEmpty) {
+    var errorMessage =
+        'Column references starting from other tables than "$tableName" are not supported. The following expressions need to be removed or modified:'
```
Comment (vlidholt) [inline packages/serverpod/lib/src/database/database_query.dart:384]: Break lines on 80 characters.

File: packages/serverpod/lib/src/database/database_query.dart [hunk 4]
```diff
@@ -349,3 +355,42 @@ _UsingQuery _usingQueryFromTableRelations(
     where: whereStatements.join(' AND '),
   );
 }
+
+void _validateTableReferences(
+  String tableName, {
+  List<Order>? orderBy,
+  Expression? where,
+}) {
+  List<String> exceptions = [];
```
Comment (vlidholt) [inline packages/serverpod/lib/src/database/database_query.dart:364]: `exceptionMessages` would be a better name, as the list does not contain exceptions.

**Review by vlidholt — APPROVED** [review]
LGTM with minor fixes.

**Pull request merged**